### ISSUE TITLE
Upload pre-release texi2html manual from CI job?

### ISSUE THREAD

**Issue by lassik** [gambit/gambit#575]
https://www.iro.umontreal.ca/~gambit/doc/gambit.html has the whole Gambit manual in one self-contained HTML file. The footer says that it's generated by [texi2html](https://www.gnu.org/software/texinfo/manual/texinfo/html_node/texi2html.html).

Could we have successful CI builds of Gambit `master` branch generate such a HTML manual and upload it to a known URL for reading in a web browser? It would be great to have up-to-the-minute HTML documentation, and probably wouldn't be hard to set up by combining existing tools.

**Comment by daviwil** [comment]
I have plans to do this with GitHub Pages :)  You can host a static website for free out of the `docs/` folder of the repo.  The url would be https://gambit.github.io/gambit.

**Comment by lassik** [comment]
True, GH Pages is a nice low effort solution for things like this.

Does it work without committing the `.html` file into Gambit's git repo?

Can we make a separate Git repo for the HTML and have Actions commit build artifacts to it?

**Comment by daviwil** [comment]
Yep, we can definitely commit to a separate repo.  Might be even better because we can more clearly delineate "stable" docs from "latest `master`" docs.  The HTML file does need to be checked in somewhere, though.

**Comment by lassik** [comment]
If a successful CI can commit it to the separate repo, that would probably be ideal.

As you imply, we can also differentiate `v4.9.3.html` vs `master.html` that way. Can the CI can check whether we're exactly on a tag, and if so, use the tag name instead of the branch name?

**Comment by daviwil** [comment]
Yep, I think that will be pretty easy with Actions.

**Comment by lassik** [comment]
> Here's a demo of generating the Gambit HTML documentation and uploading to a GitHub Pages site as part of a CI job:
> 
> https://github.com/daviwil/gambit/runs/720886829?check_suite_focus=true
> https://daviwil.com/gambit-site-demo/docs/latest/gambit.html

Great!

> Forgot to mention, it looks a little different because I've been making some tweaks to try and improve readability (responsiveness) on mobile devices.

LGTM. Thanks for the initiative. The official docs probably just use the default settings.

> I think if we create an organization site repo under the Gambit org called `gambit/gambit.github.io` then we'd be able to do this easily from the main "Gambit - CI" workflow on pushes to `master`.

I would also recommend this approach unless @feeley has an easy SFTP site ready.

**Comment by feeley** [comment]
Nice!  Well... nice except for the fact that the documentation seems straight from the 90's... nowhere as nice as the Racket docs, Mozilla docs, etc

We should explore how much we can tweak texinfo to generate a more refreshing look.

**Comment by daviwil** [comment]
Yep, I'm experimenting with that a bit to see how much control I have over the output.  Worst-case scenario, we could just post-process the HTML to take parts of it and produce a more modern output.  We might also consider breaking up the output into multiple HTML pages, the single-page file is large enough that it makes mobile browsers lag pretty hard.

**Comment by lassik** [comment]
I still aspire to scrape all Scheme implementations' manuals and serve them from doc.scheme.org. That would entail parsing TeXinfo (Guile has a parser) among a few other formats, and rendering the result as HTML.

**Comment by lassik** [comment]
> We might also consider breaking up the output into multiple HTML pages, the single-page file is large enough that it makes mobile browsers lag pretty hard.

And scrolling is a bit hard on small screens.

**Comment by lassik** [comment]
For Gambit's own website, the same parser and renderer could be used (I can turn it into a portable Scheme library) if you want to.

We should really have a Scheme equivalent to [Pandoc](https://en.wikipedia.org/wiki/Pandoc) which covers the doc formats used in the Scheme world. It allows us to be much more flexible re: output and generating indexes and the like.

**Comment by daviwil** [comment]
That'd be cool! My wheels start to turn about making a general static site generation library in Scheme that could take different types of input files (Markdown, texinfo) and convert them into HTML pages. Haunt is one example of this for Guile.

**Comment by lassik** [comment]
Great! @amirouche started a friendly competition to make static site generators last year; too bad it already ended.

If you're excited about it, I can hand off all or part of the project to you. It's meant to be part of the [Schemedoc project](https://github.com/schemedoc). We mostly talk on the [schemedoc mailing list](https://srfi-email.schemers.org/schemedoc) and the GitHub issues of that organization. Let me know if you want an invite to the org.

**Comment by lassik** [comment]
Actually my meta-level aspiration is to make an API at `api.scheme.org`. `doc.scheme.org` would pull documentation from the API and render it suitably. We have some tentative GraphQL-based stuff already at https://api.staging.scheme.org/ and the intention is to expand it a lot.

**Comment by daviwil** [comment]
Sure, I'll join! I'm certainly interested to see where I can help out with that effort.

**Comment by lassik** [comment]
That's great news for our fledgling project :) Welcome aboard.

**Comment by lassik** [comment]
@daviwil Do you want to work on some concrete tasks related to texinfo? I could try a first pass at parsing it in the coming weeks, but if you'd enjoy doing it, I can leave it to you. I'm always swamped with things to do so I don't generally need to do anything in particular :)

**Comment by daviwil** [comment]
I might be able to look at it sometime later this week.  Did you already have any code started for it?  I wonder how large of an effort it'd be to make a complete texinfo parser in Scheme, at least to the level that'd be needed for parsing Scheme implementation docs?

**Comment by feeley** [comment]
Parsing texinfo seems like a big project... perhaps theres a simpler way (look for specific markers and formatting).

**Comment by lassik** [comment]
Guile already has a texinfo parser: https://www.gnu.org/software/guile/manual/html_node/Texinfo-Processing.html IIRC there was another texinfo parser in Scheme too, but I can't find it now.

**Comment by lassik** [comment]
That parser can turn it into [SXML](http://okmij.org/ftp/Scheme/SXML.html) or something close to that, which is nice.

**Comment by lassik** [comment]
Here's a Markdown->SXML parser from Chicken: http://wiki.call-cc.org/eggref/4/lowdown

Perhaps SXML should be the common backend representation for dealing with Scheme documentation.

**Comment by daviwil** [comment]
Spent a little time today coming up with a new Gambit homepage design based on some text that @lassik sent to @feeley at some point last year:

http://daviwil.com/gambit-site-demo/

This could be the basis for the site that contains the documentation files that get pushed from CI/release builds.

**Comment by lassik** [comment]
It looks fantastic! Big thumbs up from me 👍 

Great to have someone on board who can do graphic design :) I have no skills, so I just had black text on a white background in my draft. Yours looks modern while still not looking like a copy of other programming sites - great job.

**Comment by feeley** [comment]
@daviwil The landing page looks really good!  Here are a few quick ideas/comments:

1) Add some links to the GUI-based installers for Windows and macOS, and the links: https://apps.apple.com/us/app/gambit-repl/id434534076 and https://apps.apple.com/us/app/not-emacs/id948330589

2) Add a "Try it online" button at the bottom with a link to https://feeley.github.io/gambit-in-the-browser/ (this build of Gambit with emscripten should probably be updated to be in sync with the latest release of Gambit maybe as part of CI?). I'm also considering a small development environment for Scheme based on http://codeboot.org that would be useful for students.

3) A showcase (or demos) link would be nice too

4) Link to Gambit forensics?

5) The page focuses on "traditional development workflow" specifically through C compilation.  Even though they are not quite as mature, web development and the other backends should also be mentioned (x86, ARM, JS, Python, ...).

**Comment by lassik** [comment]
> Add some links to the GUI-based installers for Windows and macOS,

Also `docker run -it schemers/gambit` for the latest release :)

`docker run -it schemers/gambit:head` is an easy way to run the pre-release version, usually not more than a couple of days old. We are exploring ways to auto-build new Docker containers as soon as commits are made to master.

> * Link to Gambit forensics?

That project is great, but the word _forensics_ has a strong association with security to most people. I.e. forensics would be about finding out whether someone has tampered with the binary or something. I got the impression Gambit Forensics is mainly concerned with performance degradations.

> * The page focuses on "traditional development workflow" specifically through C compilation.  Even though they are not quite as mature, web development and the other backends should also be mentioned (x86, ARM, JS, Python, ...).

Since Gambit constantly has research work being done, maybe there should be a separate part on the page highlighting recent activity. "Research"/"Labs"/"Upcoming"/"In the works"?

**Comment by feeley** [comment]
> > Add some links to the GUI-based installers for Windows and macOS,
> 
> Also `docker run -it schemers/gambit` for the latest release :)
> 
> `docker run -it schemers/gambit:head` is an easy way to run the pre-release version, usually not more than a couple of days old. We are exploring ways to auto-build new Docker containers as soon as commits are made to master.

Too bad it is "schemers" and not "scheme"...

> > * Link to Gambit forensics?
> 
> That project is great, but the word _forensics_ has a strong association with security to most people. I.e. forensics would be about finding out whether someone has tampered with the binary or something. I got the impression Gambit Forensics is mainly concerned with performance degradations.

True.  The name "forensics" was chosen because it means "investigation based on scientific tests and techniques", but I agree that there's an implicit aspect of "crime".  Gambit Forensics is about investigating performance anomalies and improving the performance of Gambit.

> > * The page focuses on "traditional development workflow" specifically through C compilation.  Even though they are not quite as mature, web development and the other backends should also be mentioned (x86, ARM, JS, Python, ...).
> 
> Since Gambit constantly has research work being done, maybe there should be a separate part on the page highlighting recent activity. "Research"/"Labs"/"Upcoming"/"In the works"?

Good point.

**Comment by daviwil** [comment]
When mentioning Gambit Forensics I can give the context that it's a tool for investigating performance characteristics of the Gambit system.

Working on some updates based on the other feedback!

**Comment by feeley** [comment]
We can also change the name!  Suggestions?

**Comment by daviwil** [comment]
Stream of thought suggestions:

- Gambit Tachometer
- Gambit Benchmark
- Gambit Metrics

**Comment by daviwil** [comment]
Added download links and platform info, will add some more details later:

https://daviwil.com/gambit-site-demo/

Very open to changing anything!

**Comment by lassik** [comment]
> > `docker run -it schemers/gambit`
> 
> Too bad it is "schemers" and not "scheme"...

Yes. We tried almost every trick in the book to get "scheme" but unfortunately did not succeed. "scheme" was registered in 2017 by an unknown person who doesn't have any containers or contact information: https://hub.docker.com/u/scheme And Docker, Inc. doesn't take support requests for Docker Hub so there's no way we can work it out with them.

**Comment by lassik** [comment]
> * Gambit Metrics

I like this alternative. "Benchmarks" is also nice, but I think it will lead people to expect comparisons to other language implementations besides Gambit. Then again, there were some comparisons at least to Chez in some of those charts, so maybe Benchmarks is appropriate.

**Comment by belmarca** [comment]
Good idea and the website looks nice. Can we keep discussion in one issue? Or is there a reason to have #575 and #592 ?

**Comment by feeley** [comment]
Yes I think "Gambit Metrics" is a good name.

On another note, I like the REPL on the landing page of Haskell: https://www.haskell.org

Maybe we could build a web gsi that is loaded on demand, like:

    make _gambit.js
    echo '(##repl-debug-main)' > webgsi.scm
    gsc/gsc -:= -target js -exe -o webgsi.html webgsi.scm
    open webgsi.html

**Comment by lassik** [comment]
> Can we keep discussion in one issue? Or is there a reason to have #575 and #592 ?

This issue was originally about a different thing. The website update is massive enough that it's good to have several issues about it IMHO. GitHub has a Milestone feature that I've found useful for grouping related issues together. It's implemented in a very simple and non-bureaucratic way. "Website" could be one milestone.

**Comment by lassik** [comment]
> Maybe we could build a web gsi that is loaded on demand,

You mean a gsi that everyone starts their own personal copy from their own shell, instead of one embedded on a public website?

**Comment by daviwil** [comment]
Damn, Haskell's site looks nice these days.  Clever little REPL integration too.  I'm pretty sure we could do something similar for Gambit, or at least integrate the current Gambit REPL page into the site itself (or on a subdomain like https://try.gambitscheme.org, could easily set that up as a separate GitHub Pages repo).

**Comment by APIPLM** [comment]
@lassik  You are right. We still need to a tool like [Pandoc](https://en.wikipedia.org/wiki/Pandoc) to transform different format in scheme world.The Guile have not hacked the TexInfo yet. they gives us the SXML. But I more like stexinfo. You might find people have more idea on the website [Texinfo](http://esr.ibiblio.org/?p=5211). I also are looking for why they did not kill the Texinfo.
